File: db/loaders/stops.js
'use strict';

const fs = require('fs');

const table = 'stop';

module.exports = {
  load,
  unload,
};

async function load(client, batchSize = 100) {
  const data = fs.readFileSync(`trimet/data/stops.json`);
  const parsed = JSON.parse(data);
  const stops = parsed.data;
  const columns = ['id', 'name', 'direction', 'location'];

  let values = [];

  console.info(`Loading ${stops.length} ${table}s...`);

  for (let stop of stops) {
    const [x, y] = stop.location;
    stop.location = `SRID=4326;POINT(${x} ${y})`;
    values.push(columns.map(column => stop[column]));
    if (values.length === batchSize) {
      await client.bulkInsert(table, columns, values);
      values = [];
    }
  }

  if (values.length) {
    await client.bulkInsert(table, columns, values);
  }

  const count = await client.count(table);
  console.info(`${count} ${table}s in database`);
}

async function unload(client) {
  console.info(`Deleting stops...`);
  await client.query(`DELETE FROM ${table}`);
  const count = await client.count(table);
  console.info(`${count} ${table}s in database`);
}

File: src/api/mystops/loaders/stops.py
import json
from pathlib import Path

from django.contrib.gis.geos import Point

from ..apps.stops.models import Stop


def load(path, clear, batch_size=100):
    if clear:
        print("Clearing stops...")
        Stop.objects.all().delete()

    path = Path(path)
    with path.open() as fp:
        stops = json.load(fp)["data"]

    batch = []
    total = len(stops)
    inserted = 0

    message = f"Loading {total} stops from {path}..."
    print(message, end="")

    for stop in stops:
        x, y = stop["location"]
        batch.append(
            Stop(
                stop_id=stop["id"],
                name=stop["name"],
                direction=stop["direction"],
                location=Point(x, y, srid=4326),
            )
        )
        if len(batch) == batch_size:
            Stop.objects.bulk_create(batch)
            batch.clear()
            inserted += batch_size
            percent_done = f"{inserted / total:.1%}"
            print(f"\r{message}", end=percent_done, flush=True)

    if batch:
        Stop.objects.bulk_create(batch)
        print(f"\r{message}", end="100%", flush=True)

    print(f"\r{message}", end="Done \n", flush=True)
